### PR TITLE
Telemetry server

### LINKED ISSUE #44: Telemetry GUI
I'm using this to write down my thoughts about using a Flask web server to act as the telemetry GUI.

Advantages:
- Easy to make GUIs due to wide range of bootstrap style libraries
- Will look a lot prettier than a python interface
- Flask style server allows good integration with existing python code
- Wide compatibility across devices -  could connect to the server through a tablet for example
- Potentially allow multiple simultaneous clients
- [Flask-SocketIO](https://flask-socketio.readthedocs.io/en/latest/) looks like a really simple client-server coms system

Disadvantages
- More languages required... Javascript
- Will require abstraction from rabbit-mq layer - another coms interface that needs defining

### LINKED ISSUE #49: Docker Compose/Desktop Setup
# Useful Links
- [Docker Desktop WSL 2 backend](https://docs.docker.com/docker-for-windows/wsl/)
- [Windows Subsystem for Linux Installation Guide for Windows 10](https://docs.microsoft.com/en-us/windows/wsl/install-win10)
- [Docker Compose](https://docs.docker.com/compose/install/)
- [Networking in Compose](https://docs.docker.com/compose/networking/)
- [VSCode - Use Docker Compose](https://code.visualstudio.com/docs/containers/docker-compose)
- [VSCode - Connecting to multiple containers at once](https://code.visualstudio.com/docs/remote/containers-advanced#_connecting-to-multiple-containers-at-once)

### PR DESCRIPTION
**Pull request by fsherratt**
## What's in this pull request
This pull request adds the backend and framework for a Web App style telemetry system, with communication over a UDP proxy.

Files in the `server` folder provide the web_app functionality. The `docker` folder contains the docker files required to convert the project over to docker-compose.

The following files provide the bulk of the new functionality:
- modules/udp_bridge.py
- modules/utils/async_message.py
- modules/utils/excepthook.py
- modules/utils/udp.py
- server/server.py

## How to test
Code can be tested, on any system with docker-compose installed, by running the following for in project root directory. 
```bash
docker-compose -f docker/docker-compose.gcs.yml up
```

The Flask server can then be accessed at [127.0.0.1](http://127.0.0.1). Docker-compose is handling port forwarding from `5000->80`

The server is currently set up so that messages will loopback, following the path below. 
```
socket_io -> rabbit-MQ -> UDP loopback -> rabbit-MQ -> socket_io
```
When clicking any of the web app buttons the following should be observable in your browser console
Screenshot from 2020-10-03 15-56-55

## Known Issues
- The rabbit MQ abstraction `message_broker.py` is still causing a lot of issues. This will be addressed in a future pull request (#50). For example connection timeouts exceptions if no activity happens for > 60s
- Docker-desktop doesn't forward UDP multicast packets to container :disappointed:. So you have to use a Ubuntu virtual machine to receive multicast packets from external devices. Docker-desktop works fine if using localhost to test

### CODE REVIEW

**Comment by fsherratt** [comment]
The server is currently set up so that it will route messages back at its self `socket_io 🠖 rabbit-MQ 🠖 UDP loopback 🠖 rabbit-MQ 🠖 socket_io`

File: modules/udp_bridge.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+
+
+class Message_Producer(async_message.Async_Threaded_Queue):
+    def __init__(self, host):
+        super().__init__()
+        self._host = host
+
+    def interpret_msg(self, timestanmp, msg):
+        if not isinstance(msg, data_types.Telemetry_Message):
+            return  # TODO: Err here
```
Comment (Neoxas) [inline modules/udp_bridge.py:29]: Simple todo, plz fix. Silent errors make me cry.

File: modules/udp_bridge.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        while self._running:
+            msg = self._udp.read()
+
+            if msg is None:
+                continue
+
+            try:
+                msg = pickle.loads(msg)
+            except pickle.UnpicklingError:
+                # TODO: Log that meesage couldn't be decoded
```
Comment (Neoxas) [inline modules/udp_bridge.py:103]: Add a console write? Something so we can see an error

File: modules/udp_bridge.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        self._loop_thread = threading.Thread(target=self.loop, name="UDP_Listener")
+        self._loop_thread.start()
+
+    def loop(self):
+        self._udp.openPort()
+
+        while self._running:
+            msg = self._udp.read()
+
+            if msg is None:
```
Comment (Neoxas) [inline modules/udp_bridge.py:97]: What is the point in the continue? Sure this should be an if else?

File: modules/udp_bridge.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        self._udp.openPort()
+
+    def interpret_msg(self, timestamp, msg):
+        print("UDP Broadcast: {}".format(msg), flush=True)
+
+        try:
+            msg = pickle.dumps(msg)
+        except pickle.PickleError:
+            return
+            # TODO: log error
```
Comment (Neoxas) [inline modules/udp_bridge.py:128]: Plz console.writeline

File: modules/utils/async_message.py [hunk 5]
```diff
@@ -37,8 +40,8 @@ def queue_message(self, data):
             self._message_queue.put_nowait((time.time(), data))
             self._set_message_event()
         except queue.Full:
-            #TODO: Add in logging for dropped messages
-            pass # Silently drop message for the moment
+            # TODO: Add in logging for dropped messages
+            pass  # Silently drop message for the moment
```
Comment (Neoxas) [inline modules/utils/async_message.py:44]: Plz can we add something? Surely Console.log is good enough at this point?

File: modules/utils/udp.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        self._sRead.bind(self._read_address)
+
+        self._rConnected = True
+
+    def _openWritePort(self):
+        if self._wConnected:
+            return
+
+        elif self._write_address is None:
+            # TODO: Do something more helpful
```
Comment (Neoxas) [inline modules/utils/udp.py:72]: Plz write failures or this will be a pain

File: modules/utils/udp.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    def _closeWritePort(self):
+        if self._sWrite is None:
+            return
+
+        self._sWrite.close()
+        self._wConnected = False
+
+    def read(self, b=0):
+        if not self._rConnected:
+            return
```
Comment (Neoxas) [inline modules/utils/udp.py:111]: Do we really want to return silently if we are not connected?

File: modules/utils/udp.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+            if r_list:
+                m, addr = self._sRead.recvfrom(self.buff_size)
+
+                # Write address can be learnt from incoming packet
+                if self._write_address is None:
+                    self._write_address = addr
+
+        except socket.timeout:
+            # TODO: something more useful here
```
Comment (Neoxas) [inline modules/utils/udp.py:127]: Probably should log we a dropping events

File: modules/utils/udp.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+        bytes_remaining = None
+
+        try:
+            self._write_lock.acquire()
+
+            bytes_remaining = self._sWrite.send(b)
+
+        except socket.timeout:
+            # TODO: something more useful here
```
Comment (Neoxas) [inline modules/utils/udp.py:146]: Log timeout event

File: server/server.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    excepthook,
+)
+
+
+class Server:
+    def __init__(
+        self,
+        flask_port=5000,
+        flask_host="0.0.0.0",
+        flask_secret_key="secret!",
```
Comment (Neoxas) [inline server/server.py:30]: top key. 10/10/

**Review by Neoxas — CHANGES_REQUESTED** [review]
Please for the love of god log some error. Dump them to a file or console. Just anywhere we can trace what is going wrong

**Pull request closed without merging**